Issue #3072: Add NewlineAtEndOfFile config for test and it inputs

kukreja-vikramaditya opened the pull request:
Issue: #3072

rnveach commented:
https://travis-ci.org/checkstyle/checkstyle/jobs/213574635#L382
Something is wrong with configuration, no files are being processed.
Please fix.

kukreja-vikramaditya commented:
Yes i think I added  **/* in excludes. Sending a new PR

kukreja-vikramaditya commented:
Some files like `package-info.java` or `java.header` or `import-control_complete.xml`  are marked are ERROR for this test.. Is that acceptable, or just files of the sort `Input.*.java` are to be checked

kukreja-vikramaditya commented:
The current PR has a bug in excludes property of the config xml. On fixing this and running `mvn clean verify` causes build to fail since there files without an eol. Should i fix the error for all those files and push again? I tried setting​ failOnViolation to false also but to no avail. Please advice.

rnveach commented:
> Some files like package-info.java or java.header or import-control_complete.xml (all inside test/resources

IMO, all these file types that are text files in our resources should be validated.

> Should i fix the error for all those files and push again? 

Yes, errors must be fixed. We can't merge a PR if CIs fail.
Be sure to not make to changes to inputs used by `NewlineAtEndOfFileCheckTest`. Any errors for those files must remain. You will need to add suppressions to ignore their violations and pass CI.

kukreja-vikramaditya commented:
```C:\projects\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\InputNewlineLfAtEndOfFile.java:0: File does not end with a newline. [NewlineAtEndOfFile]
[checkstyle] [ERROR] C:\projects\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\grammars\InputEmbeddedNullChar.java:0: File does not end with a newline. [NewlineAtEndOfFile]
``` Fails in windows system. LF is obvious for `InputNewlineLfAtEndOfFile`. Is it the same for the other file as well?

rnveach commented:
> InputNewlineLfAtEndOfFile.java:0

Check out `NewlineAtEndOfFileCheckTest`, this is an input file to the test. It can't be changed.
We will need a suppression for these files.

> InputEmbeddedNullChar.java

I'm not sure why this is failing. I see a `\n` in linux on the end and travis is passing. I wonder if it is because of the `\0`.
Do you have a windows machine to look into it?

kukreja-vikramaditya commented:
I used a hex editor to see the content of `InputEmbeddedNullChar.java`. It ends with `0A` . I have added `0D` before it and will update the PR. Hopefully this should resolve the error on a windows machine. Otherwise this can be added to the suppressions list.
Also added all inputs that are used by `NewlineAtEndOfFileCheckTest`

kukreja-vikramaditya commented:
Do you think that the javadoc and grammar input folders need to be added to suppresion list ? As adding input files one by one in the `input-eol-verify.xml` may be difficult to track in the future?

eg.
```xml
test/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/htmlTags/*,
test/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/javadocTags/*,
```
Many of the input files in these folders cannot have in eol at the end (tests fail if we attempt to do so). I have already updated all files that needed and could accomodate an eol.

rnveach commented:
> InputEmbeddedNullChar.java. It ends with 0A .

It ends with the correct byte, which is why linux passes. The question is why Windows isn't working the same since we use `autocrlf=true`.
I examined the case on my windows machine and I can confirm git is not transforming it to `\r\n` and leaving it as `\n`.
The reason is git and github sees the file as binary since it has the `\0` in it. It won't convert binary files.
Github: https://github.com/checkstyle/checkstyle/pull/4075/commits/01ee7c2618e0a7ba812a2437ff0811f3422ae885#diff-650cb5587add6caa98601b2adc4d8fff (says `BIN` on side)
Git: http://stackoverflow.com/questions/6119956/how-to-determine-if-git-handles-a-file-as-binary-or-as-text

> I have added 0D before it and will update the PR. Hopefully this should resolve the error on a windows machine. Otherwise this can be added to the suppressions list.

Travis should have failed, imo. Default line separator is `SYSTEM` which is `\n` on linux and `\r\n` on windows. This might be related to Issue #4074.

Please undo the change to this file and add it as a separation.
Please wait for me to review the rest of the PR.

kukreja-vikramaditya commented:
I'll keep the changes ready. Let me know when to push after your review of this issue and #4074

kukreja-vikramaditya commented:
@rnveach 
All changes made.

rnveach commented on config/resource_suppressions.xml:
```diff
@@ -0,0 +1,12 @@
+<suppressions>
+    
+      <!-- Suppressions inside test and it folders -->
+      <suppress checks="NewlineAtEndOfFile" files="(src[\\/]it[\\/].*rule731selfexplanatory[\\/]InputJavadocMethodCheck\.java|src[\\/]test[\\/].*Newline.*AtEndOfFile\.java|src[\\/]test[\\/].*astprinter[\\/]InputJavadocComment\.javadoc|src[\\/]test[\\/].*astprinter[\\/]InputJavadocWithError\.javadoc|src[\\/]test[\\/].*checks[\\/]InputEmptyFile\.txt|src[\\/]test[\\/].*grammars[\\/]javadoc[\\/]htmlT ...
```
This is too hard to read and maintain as one line string. What if we accidentally screw expression somewhere.
Please split it up. I am fine with one file/directory per line.

kukreja-vikramaditya replied:
Should I add reasons for all suppressions or just a general message at the top saying smth like this "UTs that use these files cannot work with a new line at the end". Imho first approach may be a good practice for future suppressions?

rnveach replied:
> Should I add reasons for all individual suppressions or just a general message at the top

General message if all of them have identical reasons.
`Newline.*AtEndOfFile\.java` can have its own unique reason which is because the check is dependent on them for violations.

> Imho first approach may be a good practice for future suppressions?

Explain what you mean by `first approach`.
If you mean all on one line in one big expression, then no for the reasons I stated.
If you mean in the ant XML file, then no, because it limits the type of suppressions we can do by check.

kukreja-vikramaditya replied:
>If you mean all on one line in one big expression, then no for the reasons I stated.

The reason for this is clear to me. 

No i meant that comment for all similar types of suppressions. Eg
```xml
<!-- Input files for NewlineAtEndOfFileCheckTest-->
      <suppress checks="NewlineAtEndOfFile" files="src[\\/]test[\\/].*Newline.*AtEndOfFile\.java"/>

<!-- Many files in these folder require no newline at the end and are suppressed by default -->
      <suppress checks="NewlineAtEndOfFile" files="src[\\/]test[\\/].*grammars[\\/]javadoc[\\/]htmlTags[\\/].*"/>
      <suppress checks="NewlineAtEndOfFile" files="src[\\/]test[\\/].*grammars[\\/]javadoc[\\/]javadocTags[\\/].*"/>

<!--CI fails on windows system for this file -->
      <suppress checks="NewlineAtEndOfFile" files="src[\\/]test[\\/].*grammars[\\/]InputEmbeddedNullChar\.java"/>

<!-- Some specific javadoc input files fail tests when they end with new line-->
      <suppress checks="NewlineAtEndOfFile" files="src[\\/]test[\\/].*grammars[\\/]javadoc[\\/]InputLeadingAsterisks\.txt"/>
      <suppress checks="NewlineAtEndOfFile" files="src[\\/]test[\\/].*grammars[\\/]comments[\\/]InputFullOfSinglelineComments\.java"/>
      <suppress checks="NewlineAtEndOfFile" files="src[\\/]test[\\/].*astprinter[\\/]InputJavadocComment\.javadoc"/>
      <suppress checks="NewlineAtEndOfFile" files="src[\\/]test[\\/].*astprinter[\\/]InputJavadocWithError\.javadoc"/>
      <suppress checks="NewlineAtEndOfFile" files="src[\\/]it[\\/].*rule731selfexplanatory[\\/]InputJavadocMethodCheck\.java"/>

<!--Empty file cannot have any data  -->
      <suppress checks="NewlineAtEndOfFile" files="src[\\/]test[\\/].*checks[\\/]InputEmptyFile\.txt"/>
```

rnveach replied:
Yes, I am fine with one comment for all similar types of suppressions. This is how our main suppression file is too.

rnveach commented on config/resource_suppressions.xml:
```diff
@@ -0,0 +1,12 @@
+
+<suppressions>
+    
+      <!-- Suppressions inside test and it folders -->
```
This whole configuration is for test and it folders. So comment is redundant.
Make comment specify it is for why we can't add newline to end of files or none at all.

kukreja-vikramaditya commented:
On a different note, as safe measure i have kept the starting regexp as `src[\\/]test[\\/]` because I was getting some files from target folder too during detection

rnveach commented:
> getting some files from target folder 

https://github.com/checkstyle/checkstyle/pull/4075/commits/d4bd6cb493920c8259130e138666e4f5677def17#diff-2e6859b1ab103e7786fee59edc13848cR27
This is an issue with too lax of an includes or a missing exclude. `target` folder should not be scanned for anything.

kukreja-vikramaditya commented:
Are you saying `src/` should be appended before `it` and `test`. That would reduce the regexp in the suppressions list. Will include in updated PR. Please review comment also https://github.com/checkstyle/checkstyle/pull/4075#discussion_r107781054

rnveach commented:
> Are you saying src/ should be appended before it and test. 

That is fine if it prevents scanning `target`.

> Please review comment also #4075 (comment)

I did. I am fine with your example.

kukreja-vikramaditya commented:
@rnveach 
Will add the remaining changes from your review and send updated PR after @romani 's review.

romani commented on config/input-resource-verify.xml:
```diff
@@ -0,0 +1,44 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I do not like ant solution, for this case, it is over complication.
We still use ant only because it allow us to run checkstyle base on generated class files (before final jar is created).
As you can see it is ugly by configuration to make that hack.

Your case is simpler, you do not need bleeding edge version of this Check, it very simple and unlikely have a problem, but even it got a problem, it has enough inputs during main validation.

Please backup your solution and try to use validation as by maven plugin - https://github.com/checkstyle/checkstyle/blob/master/pom.xml#L457 , by default maven put to validation only src/main folder, you need to configure it to scan xdoc, site, .ci, config, ...... folders.

rnveach replied:
@romani @kukreja-vikramaditya After confirming with cache file from main config run.

`xdoc`, `site, is controlled by main config, so it shouldn't be scanned by "resource" config.
`.ci` folder is ok to add for this PR.

> config

@romani I will create a new issue for this, but if `config` file is scanned for violations and a violation is found, it will invalidate entire cachefile even if config is never changed.

romani commented on config/resource_suppressions.xml:
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0"?>
```
please name it "newline_at_end_of_file_suppressions.xml"

rnveach commented on config/input-resource-verify.xml:
```diff
@@ -0,0 +1,44 @@
+             />
+
+    <property name="checkstyle.pattern.todo" value="NOTHingWillMatCH_-"/>
+    <property name="check.config" location="config/checkstyle_resource_checks.xml"/>
```
@romani I specifically wanted it to be a general name as I think we should add other checks to the configuration besides `NewlineAtEndOfFile`.

The 2 I would like to add is: package name should be same as folder location, class name should be the same as file name.
Some others that have confused new users are tabs and trailing spaces.

rnveach replied:
@kukreja-vikramaditya After talking with @romani , he agrees about extending this for more checks in the future.
`OuterTypeFilenameCheck` requires `TreeWalker` which we can't run on non-compilable files, so we will have to examine this in new issue outside this PR.

For now, we will make a generic name for file so it makes sense if and when we extend it.
Config file will be named `checkstyle_non_main_files_checks.xml`. All other names required (suppressions and cachefile) should also be named around the config name.

rnveach commented:
@kukreja-vikramaditya As you are making changes, keep this PR rebased on latest master to avoid conflicts.

rnveach commented on config/input-resource-verify.xml:
```diff
@@ -0,0 +1,44 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<project name="input-validate" default="execute">
```
Any folders this config scans, it should pickup all files in them.
Examining cachefile generated shows what files it is finding.

kukreja-vikramaditya commented:
@romani
@rnveach 
I have made the changes as per your reviews. Local build passed. But  I am unable to understand Codecov and TeamCity build fails and why they are occuring

kukreja-vikramaditya commented:
From inspecting the cache file for this change, I noticed that main resources were also checked...

This is the configuration settings [here](https://github.com/checkstyle/checkstyle/pull/4075/commits/920ff187a0cecf1f0769affb77f2ef1c7aa97024#diff-600376dffeb79835ede4a0b285078036)..please advice what should be the value of 

```xml
              <includeResources>true</includeResources>
              <includeTestResources>true</includeTestResources>
```

rnveach commented:
> But I am unable to understand Codecov and TeamCity build fails

TC can be ignored for now as we are experiencing a failure with them. See Issue #3616 .
`codecov` might be something new as we have hit this a few times with recent changes. Unless @romani has something, I would ignore it for now as travis' code coverage passes.

kukreja-vikramaditya commented:
>Please confirm and make the change.

Yes that was the culprit. Have made the changes and pushed again

rnveach commented on config/checkstyle_non_main_files_suppressions.xml:
```diff
@@ -0,0 +1,29 @@
+<suppressions>
+    
+    <!-- Input files for NewlineAtEndOfFileCheckTest, intentional no new line at the end. -->
+    <suppress checks="NewlineAtEndOfFile" files="test.*Newline.*AtEndOfFile\.java"/>
```
`test` -> `[\\/]test[\\/]`.
Append `[\\/]` to the beginning of `Newline`.
Without directory slashes, name is too general imo.

Do similar changes to other suppressions.

rnveach commented on config/checkstyle_non_main_files_checks.xml:
```diff
@@ -0,0 +1,18 @@
+    <property name="charset" value="UTF-8"/>
+    
+    <!-- Hard coded until https://issues.apache.org/jira/browse/MCHECKSTYLE-332 is released. -->
+    <property name="cacheFile" value="target/checkstyle_non_main_files_cache"/>
```
`checkstyle_non_main_files_cache` -> `cache_non_main_files`

rnveach commented:
@kukreja-vikramaditya @romani ping

I think we should try and finish this one first since it is the biggest and will affect most other PRs.
Please rebase off latest master since this has been sitting for some time so we make sure we aren't missing anything else on a merge.

kukreja-vikramaditya commented:
I'll rebase and update the changes.

kukreja-vikramaditya commented:
@romani
@rnveach 
Please do final review

romani commented:
@kukreja-vikramaditya , please move fixes for all files (your first commit) to separate PR - I will merge it asap.
Configuration update still require additional tuning.

romani commented on pom.xml:
```diff
@@ -472,22 +472,42 @@
+              <violationSeverity>error</violationSeverity>
+              <propertyExpansion>project.basedir=${project.basedir}</propertyExpansion>
+              <sourceDirectory>${project.basedir}/src</sourceDirectory>
+              <excludes>**/it/resources/**/*,**/test/resources/**/*,**/test/resources-noncompilable/**/*</excludes>
```
You configured Checker to process all files.
But you do not process all files from repository, you do this only under "src" folder of repo.
please extend config to process all files in git repository (yml, sh, .......) except for ".git" folder.

Vladlis replied:
> except for ".git" folder

maybe also for everything mentioned in .gitignore?

romani replied:
yes, good catch

kukreja-vikramaditya replied:
Thanks for the heads up. Will make the changes

romani replied:
please do not reply to us as "Will" , You will do it for sure. just reply to us when it is  "done".

romani commented on pom.xml:
```diff
@@ -472,22 +472,42 @@
+              <propertyExpansion>project.basedir=${project.basedir}</propertyExpansion>
+              <sourceDirectory>${project.basedir}</sourceDirectory>
+              <includes>**/*</includes>
+              <excludes>src/main/**/*,src/it/java/**/*,src/test/java/**/*,src/site/resources/images/**/*,.git/**/*,.idea/**/*,target/**/*</excludes>
```
remove "src/main/**/*,src/it/java/**/*,src/test/java/**/*," to shorten config, one more launch over this folders will not be a performance problem for validation.

please add all excludes from https://github.com/checkstyle/checkstyle/blob/master/.gitignore

kukreja-vikramaditya replied:
```xml
            <excludes>
                <exclude>src/main/**/*</exclude>
                <exclude>src/it/java/**/*</exclude>
                <exclude>src/test/java/**/*</exclude>
                <exclude>src/site/resources/images/**/*</exclude>
                <exclude>target/**/*<exclude>
                <exclude>bin/**/*</exclude>
                <exclude>.git/**/*</exclude>
                <exclude>.idea/**/*</exclude>
                <exclude>.settings/**/*</exclude>
                <exclude>.checkstyle/**/*<exclude>
                <exclude>.DS_Store/**/*</exclude>
                <exclude>.nondex/**/*</exclude>
                <exclude>.pmd/**/*</exclude>
                <exclude>.fbExcludeFilterFile/**/*<exclude>
                <exclude>.externalToolBuilders/**/*</exclude>
                <exclude>.pmdruleset.xml</exclude>
                <exclude>checkstyle.iml</exclude>
                <exclude>nbactions.xml</exclude>
                <exclude>nb-configuration.xml</exclude>
           </excludes>
```

romani replied:
good. please test it

kukreja-vikramaditya replied:
Also let me know what exclude should be written for 
```xml
# Java Virtual machine crash logs
hs_err_pid*
replay_pid*
```

Vladlis replied:
`checkstyle.iml` isn't right, actually name of an .iml file depends on how you call your project in IDEA
